From fc5bd7ee79804beccf67ecd22269408afaff6aaa Mon Sep 17 00:00:00 2001
From: Guan Xuetao <gxt@mprc.pku.edu.cn>
Date: Mon, 6 Aug 2012 23:07:17 +0800
Subject: [PATCH 028/247] unicore64: Add singlestep processing

Signed-off-by: Guan Xuetao <gxt@mprc.pku.edu.cn>
---
 target-unicore64/translate.c |   23 ++++++++++++++++++++++-
 1 file changed, 22 insertions(+), 1 deletion(-)

diff --git a/target-unicore64/translate.c b/target-unicore64/translate.c
index babd325..4b87509 100644
--- a/target-unicore64/translate.c
+++ b/target-unicore64/translate.c
@@ -62,6 +62,21 @@ typedef struct DisasContext {
     struct TranslationBlock *dc_tb;
 } DisasContext;
 
+static inline void gen_goto_tb(DisasContext *s, int n, target_ulong dest)
+{
+    TranslationBlock *tb;
+
+    tb = s->dc_tb;
+    if ((tb->pc & TARGET_PAGE_MASK) == (dest & TARGET_PAGE_MASK)) {
+        tcg_gen_goto_tb(n);
+        tcg_gen_movi_i64(cpu_R[31], dest);
+        tcg_gen_exit_tb((tcg_target_long)tb + n);
+    } else {
+        tcg_gen_movi_i64(cpu_R[31], dest);
+        tcg_gen_exit_tb(0);
+    }
+}
+
 static void do_datap(CPUUniCore64State *env, DisasContext *s, uint32_t insn)
 {
 }
@@ -194,7 +209,13 @@ done_disas_loop:
          *  - Hardware watchpoints.
          * Hardware breakpoints have already been handled and skip this code.
          */
-        UNHANDLED_FLOW;
+        switch (dc->dc_jmp) {
+        case DISAS_NEXT:
+            gen_goto_tb(dc, 1, dc->dc_pc);
+            break;
+        default:
+            UNHANDLED_FLOW;
+        }
     }
 
     gen_icount_end(tb, num_insns);
-- 
1.7.9.5

