From 98f3a9f57e22cdffa914c37fc09d64c282758108 Mon Sep 17 00:00:00 2001
From: Guan Xuetao <gxt@mprc.pku.edu.cn>
Date: Wed, 8 Aug 2012 17:36:49 +0800
Subject: [PATCH 045/247] unicore64: Add insn DNOT NOT handlers

Signed-off-by: Guan Xuetao <gxt@mprc.pku.edu.cn>
---
 target-unicore64/translate.c |   12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/target-unicore64/translate.c b/target-unicore64/translate.c
index bd3b4ce..49e0c41 100644
--- a/target-unicore64/translate.c
+++ b/target-unicore64/translate.c
@@ -244,6 +244,18 @@ static void do_datap(CPUUniCore64State *env, DisasContext *s, uint32_t insn)
         }
         tcg_temp_free_i64(t_op2_64);
         break;
+    case 0x0f: /* insn NOT DNOT */
+        ILLEGAL_INSN(UCOP_REG_S1);
+
+        if (UCOP_SET(22)) { /* insn DNOT */
+            tcg_gen_not_i64(cpu_R[UCOP_REG_D], t_op2_64);
+        } else { /* insn NOT */
+            tcg_gen_not_i32(t_op2_32, t_op2_32);
+            tcg_gen_extu_i32_i64(cpu_R[UCOP_REG_D], t_op2_32);
+            tcg_temp_free_i32(t_op2_32);
+        }
+        tcg_temp_free_i64(t_op2_64);
+        break;
     default:
         ILLEGAL_INSN(true);
     }
-- 
1.7.9.5

