From c76c934cfe1b07460a145f6fc6bf32de511d0501 Mon Sep 17 00:00:00 2001
From: Guan Xuetao <gxt@mprc.pku.edu.cn>
Date: Thu, 9 Aug 2012 21:35:15 +0800
Subject: [PATCH 062/247] unicore64: Add ldst handling when base reg is pc

Signed-off-by: Guan Xuetao <gxt@mprc.pku.edu.cn>
---
 target-unicore64/translate.c |    9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/target-unicore64/translate.c b/target-unicore64/translate.c
index 89c68e6..1f9850d 100644
--- a/target-unicore64/translate.c
+++ b/target-unicore64/translate.c
@@ -549,14 +549,17 @@ static void do_ldst(CPUUniCore64State *env, DisasContext *s, uint32_t insn)
     ILLEGAL_INSN((insn >> 29) != 3); /* ONLY 011 is handled */
     ILLEGAL_INSN(!UCOP_LDST_BHD); /* prefetch or sync? */
     ILLEGAL_INSN(UCOP_REG_D == 31);
-    ILLEGAL_INSN(UCOP_REG_S1 == 31);
 
     t_addr = tcg_temp_new_i64();
     t_op2_64 = tcg_temp_new_i64();
     t_rd_64 = tcg_temp_new_i64();
 
     /* Prepare base address */
-    tcg_gen_mov_i64(t_addr, cpu_R[UCOP_REG_S1]);
+    if (UCOP_REG_S1==31) {
+        tcg_gen_movi_i64(t_addr, s->dc_pc - 4);
+    } else {
+        tcg_gen_mov_i64(t_addr, cpu_R[UCOP_REG_S1]);
+    }
 
     /* Prepare op2 */
     if (UCOP_SET(21)) { /* reg or imm */
@@ -636,6 +639,8 @@ static void do_ldst(CPUUniCore64State *env, DisasContext *s, uint32_t insn)
     }
 
     if (!UCOP_SET(27) || UCOP_SET(26)) { /* post || writeback */
+        ILLEGAL_INSN(UCOP_REG_S1 == 31);
+
         tcg_gen_mov_i64(cpu_R[UCOP_REG_S1], t_addr);
     }
 
-- 
1.7.9.5

