From 2de9aed6ac15466e6d53b0dd5ced727948652b7e Mon Sep 17 00:00:00 2001
From: Guan Xuetao <gxt@mprc.pku.edu.cn>
Date: Sat, 11 Aug 2012 10:18:57 +0800
Subject: [PATCH 082/247] unicore64: Change logic in helper_add/sub_cc

Signed-off-by: Guan Xuetao <gxt@mprc.pku.edu.cn>
---
 target-unicore64/helper.c |   12 ++++--------
 1 file changed, 4 insertions(+), 8 deletions(-)

diff --git a/target-unicore64/helper.c b/target-unicore64/helper.c
index b4c15b4..25a4802 100644
--- a/target-unicore64/helper.c
+++ b/target-unicore64/helper.c
@@ -97,8 +97,7 @@ uint32_t HELPER(sub_cc_i32)(uint32_t a, uint32_t b)
 {
     uint32_t result;
     result = a - b;
-    env->NF = result >> 31;
-    env->ZF = (result == 0);
+    env->NF = env->ZF = (int32_t)result;
     env->CF = a >= b;
     env->VF = ((a ^ b) & (a ^ result)) >> 31;
     return result;
@@ -108,8 +107,7 @@ uint64_t HELPER(sub_cc_i64)(uint64_t a, uint64_t b)
 {
     uint64_t result;
     result = a - b;
-    env->NF = result >> 63;
-    env->ZF = (result == 0);
+    env->NF = env->ZF = result;
     env->CF = a >= b;
     env->VF = ((a ^ b) & (a ^ result)) >> 63;
     return result;
@@ -119,8 +117,7 @@ uint32_t HELPER(add_cc_i32)(uint32_t a, uint32_t b)
 {
     uint32_t result;
     result = a + b;
-    env->NF = result >> 31;
-    env->ZF = (result == 0);
+    env->NF = env->ZF = (int32_t)result;
     env->CF = result < a;
     env->VF = ((a ^ b ^ -1) & (a ^ result)) >> 31;
     return result;
@@ -130,8 +127,7 @@ uint64_t HELPER(add_cc_i64)(uint64_t a, uint64_t b)
 {
     uint64_t result;
     result = a + b;
-    env->NF = result >> 63;
-    env->ZF = (result == 0);
+    env->NF = env->ZF = result;
     env->CF = result < a;
     env->VF = ((a ^ b ^ -1) & (a ^ result)) >> 63;
     return result;
-- 
1.7.9.5

