From 037bace9844dd17fe425fb46dec3c7b4a2a1006e Mon Sep 17 00:00:00 2001
From: Guan Xuetao <gxt@mprc.pku.edu.cn>
Date: Sat, 11 Aug 2012 11:30:08 +0800
Subject: [PATCH 084/247] unicore64-bugfix: Adjust VF setting in add_cc and
 sub_cc

Signed-off-by: Guan Xuetao <gxt@mprc.pku.edu.cn>
---
 target-unicore64/helper.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/target-unicore64/helper.c b/target-unicore64/helper.c
index 25a4802..f6b2cfc 100644
--- a/target-unicore64/helper.c
+++ b/target-unicore64/helper.c
@@ -119,7 +119,7 @@ uint32_t HELPER(add_cc_i32)(uint32_t a, uint32_t b)
     result = a + b;
     env->NF = env->ZF = (int32_t)result;
     env->CF = result < a;
-    env->VF = ((a ^ b ^ -1) & (a ^ result)) >> 31;
+    env->VF = (int32_t)((a ^ b ^ -1) & (a ^ result));
     return result;
 }
 
@@ -129,7 +129,7 @@ uint64_t HELPER(add_cc_i64)(uint64_t a, uint64_t b)
     result = a + b;
     env->NF = env->ZF = result;
     env->CF = result < a;
-    env->VF = ((a ^ b ^ -1) & (a ^ result)) >> 63;
+    env->VF = (a ^ b ^ -1) & (a ^ result);
     return result;
 }
 
-- 
1.7.9.5

