From ec09863b2aa8519575ced5d20bd0a78189659a17 Mon Sep 17 00:00:00 2001
From: Liu Guoli <liuguoli@mprc.pku.edu.cn>
Date: Mon, 13 Aug 2012 15:22:37 +0800
Subject: [PATCH 100/247] unicore64-softmmu: add macros for mmu

Signed-off-by: Liu Guoli <liuguoli@mprc.pku.edu.cn>
---
 target-unicore64/softmmu.c |   14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/target-unicore64/softmmu.c b/target-unicore64/softmmu.c
index cda5078..0c1a22c 100644
--- a/target-unicore64/softmmu.c
+++ b/target-unicore64/softmmu.c
@@ -16,6 +16,20 @@
 #error This file only exist under softmmu circumstance
 #endif
 
+#undef DEBUG_UC64
+
+#ifdef DEBUG_UC64
+#define DPRINTF(fmt, ...) printf("%s: " fmt , __func__, ## __VA_ARGS__)
+#else
+#define DPRINTF(fmt, ...) do {} while (0)
+#endif
+
+#define SUPERPAGE_SIZE             (1 << 20)
+#define UC64_PAGETABLE_READ        (1 << 7)
+#define UC64_PAGETABLE_WRITE       (1 << 6)
+#define UC64_PAGETABLE_EXEC        (1 << 5)
+#define UC64_PAGETABLE_EXIST       (1 << 2)
+
 #define MMUSUFFIX _mmu
 
 #define SHIFT 0
-- 
1.7.9.5

