From 04d3d67aee7eeb920f33dade7c5ccad0d8575e19 Mon Sep 17 00:00:00 2001
From: Guan Xuetao <gxt@mprc.pku.edu.cn>
Date: Mon, 13 Aug 2012 20:43:07 +0800
Subject: [PATCH 105/247] unicore64-bugfix: Correct VF setting in helpers

Signed-off-by: Guan Xuetao <gxt@mprc.pku.edu.cn>
---
 target-unicore64/helper.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/target-unicore64/helper.c b/target-unicore64/helper.c
index e14a29b..f584b9d 100644
--- a/target-unicore64/helper.c
+++ b/target-unicore64/helper.c
@@ -100,7 +100,7 @@ uint32_t HELPER(sub_cc_i32)(uint32_t a, uint32_t b)
     result = a - b;
     env->NF = env->ZF = (int32_t)result;
     env->CF = a >= b;
-    env->VF = ((a ^ b) & (a ^ result));
+    env->VF = (int32_t)((a ^ b) & (a ^ result));
     return result;
 }
 
@@ -144,7 +144,7 @@ uint32_t HELPER(sbc_cc_i32)(uint32_t a, uint32_t b)
         result = a - b;
         env->CF = a >= b;
     }
-    env->VF = ((a ^ b) & (a ^ result));
+    env->VF = (int32_t)((a ^ b) & (a ^ result));
     env->NF = env->ZF = (int32_t)result;
     return result;
 }
-- 
1.7.9.5

