From 97c7e0d296555d8945a4531fede06be82d12e0f4 Mon Sep 17 00:00:00 2001
From: Guan Xuetao <gxt@mprc.pku.edu.cn>
Date: Mon, 13 Aug 2012 21:43:11 +0800
Subject: [PATCH 109/247] unicore64-softmmu-bugfix: Correct entry address for
 zImage

Signed-off-by: Guan Xuetao <gxt@mprc.pku.edu.cn>
---
 target-unicore64/cpu.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/target-unicore64/cpu.c b/target-unicore64/cpu.c
index e9eb83a..d376ddb 100644
--- a/target-unicore64/cpu.c
+++ b/target-unicore64/cpu.c
@@ -21,7 +21,7 @@ static void uc64_unicore3_cpu_initfn(Object *obj)
 
     env->cp0.c0_cpuid = 0x4d110863;
     env->uncached_asr = ASR_MODE_PRIV;
-    env->regs[31] = 0x300000;
+    env->regs[31] = 0x03000000;
 }
 
 static void uc64_any_cpu_initfn(Object *obj)
-- 
1.7.9.5

