From 5a15f496c6458f6d4602697e99493c7746831a53 Mon Sep 17 00:00:00 2001
From: Guan Xuetao <gxt@mprc.pku.edu.cn>
Date: Wed, 15 Aug 2012 01:28:41 +0800
Subject: [PATCH 123/247] unicore64-softmmu: Add cp0_c12 get handler

Signed-off-by: Guan Xuetao <gxt@mprc.pku.edu.cn>
---
 target-unicore64/helper.c |    5 +++++
 1 file changed, 5 insertions(+)

diff --git a/target-unicore64/helper.c b/target-unicore64/helper.c
index fa8a317..d0b9e98 100644
--- a/target-unicore64/helper.c
+++ b/target-unicore64/helper.c
@@ -243,6 +243,11 @@ uint64_t helper_cp0_get(CPUUniCore64State *env, uint64_t creg,
             return env->cp0.c1_sys;
         }
         break;
+    case 12:
+        if (cop < 8) {
+            return env->cp0.c12_sysu[cop];
+        }
+        break;
     }
     DPRINTF("Wrong register (%" PRIx64 ") or wrong operation (%" PRIx64
             ") in %s!\n", creg, cop, __func__);
-- 
1.7.9.5

