From feef4d4a75ed422c1bcba8c6278ee9aa860836cb Mon Sep 17 00:00:00 2001
From: Meng Xiangzeng <mengxiangzeng@mprc.pku.edu.cn>
Date: Tue, 14 Aug 2012 23:31:24 +0800
Subject: [PATCH 124/247] unicore64: Add insn ERET handler

Signed-off-by: Meng Xiangzeng <mengxiangzeng@mprc.pku.edu.cn>
---
 target-unicore64/translate.c |   12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/target-unicore64/translate.c b/target-unicore64/translate.c
index 1b22544..9454363 100644
--- a/target-unicore64/translate.c
+++ b/target-unicore64/translate.c
@@ -1134,6 +1134,7 @@ static void do_llsc(CPUUniCore64State *env, DisasContext *s, uint32_t insn)
 static void do_branch(CPUUniCore64State *env, DisasContext *s, uint32_t insn)
 {
     target_ulong t_addr;
+    TCGv_i64 tmp;
 
     if (UCOP_SET(28)) { /* link */
         /* r30 <- next_insn */
@@ -1156,7 +1157,16 @@ static void do_branch(CPUUniCore64State *env, DisasContext *s, uint32_t insn)
                 s->dc_jmp = DISAS_JUMP;
                 break;
             case 0x00c00000:
-                /* ERET instruction: r31 <- r30, ASR <- BSR */
+                /* ERET instruction: r31 <- p0.c4_itrapaddr, ASR <- BSR */
+                tmp = tcg_temp_new_i64();
+                gen_load_cpu_field(cpu_R[31], cp0.c4_itrapaddr);
+                gen_load_cpu_field(tmp, bsr);
+                gen_helper_asr_write(tmp);
+                gen_load_cpu_field(tmp, bfr);
+                gen_helper_afr_write(tmp);
+                tcg_temp_free_i64(tmp);
+                s->dc_jmp = DISAS_JUMP;
+                break;
             default:
                 ILLEGAL_INSN(true);
             }
-- 
1.7.9.5

