From 1bc3c3bf6f2bb9cfaea56dce094b75d8504931ac Mon Sep 17 00:00:00 2001
From: Guan Xuetao <gxt@mprc.pku.edu.cn>
Date: Wed, 15 Aug 2012 18:10:11 +0800
Subject: [PATCH 130/247] unicore64: Adjust syscall_nr.h to match linux kernel

Signed-off-by: Guan Xuetao <gxt@mprc.pku.edu.cn>
---
 linux-user/unicore64/syscall_nr.h |   49 ++++++++++++++-----------------------
 1 file changed, 18 insertions(+), 31 deletions(-)

diff --git a/linux-user/unicore64/syscall_nr.h b/linux-user/unicore64/syscall_nr.h
index 2f778db..1c69eec 100644
--- a/linux-user/unicore64/syscall_nr.h
+++ b/linux-user/unicore64/syscall_nr.h
@@ -28,7 +28,7 @@
 #define TARGET_NR_epoll_pwait 22
 #define TARGET_NR_dup 23
 #define TARGET_NR_dup3 24
-#define TARGET_NR_fcntl64 25
+#define TARGET_NR_fcntl 25
 #define TARGET_NR_inotify_init1 26
 #define TARGET_NR_inotify_add_watch 27
 #define TARGET_NR_inotify_rm_watch 28
@@ -46,10 +46,10 @@
 #define TARGET_NR_mount 40
 #define TARGET_NR_pivot_root 41
 #define TARGET_NR_nfsservctl 42
-#define TARGET_NR_statfs64 43
-#define TARGET_NR_fstatfs64 44
-#define TARGET_NR_truncate64 45
-#define TARGET_NR_ftruncate64 46
+#define TARGET_NR_statfs 43
+#define TARGET_NR_fstatfs 44
+#define TARGET_NR_truncate 45
+#define TARGET_NR_ftruncate 46
 #define TARGET_NR_fallocate 47
 #define TARGET_NR_faccessat 48
 #define TARGET_NR_chdir 49
@@ -62,10 +62,10 @@
 #define TARGET_NR_openat 56
 #define TARGET_NR_close 57
 #define TARGET_NR_vhangup 58
-#define TARGET_NR_pipe2 59
+#define TARGET_NR_pipe 59 /* pipe2 */
 #define TARGET_NR_quotactl 60
 #define TARGET_NR_getdents64 61
-#define TARGET_NR_llseek 62
+#define TARGET_NR_lseek 62
 #define TARGET_NR_read 63
 #define TARGET_NR_write 64
 #define TARGET_NR_readv 65
@@ -74,7 +74,7 @@
 #define TARGET_NR_pwrite64 68
 #define TARGET_NR_preadv 69
 #define TARGET_NR_pwritev 70
-#define TARGET_NR_sendfile64 71
+#define TARGET_NR_sendfile 71
 #define TARGET_NR_pselect6 72
 #define TARGET_NR_ppoll 73
 #define TARGET_NR_signalfd4 74
@@ -82,8 +82,8 @@
 #define TARGET_NR_splice 76
 #define TARGET_NR_tee 77
 #define TARGET_NR_readlinkat 78
-#define TARGET_NR_fstatat64 79
-#define TARGET_NR_fstat64 80
+#define TARGET_NR_fstatat 79
+#define TARGET_NR_fstat 80
 #define TARGET_NR_sync 81
 #define TARGET_NR_fsync 82
 #define TARGET_NR_fdatasync 83
@@ -280,27 +280,10 @@
 #define TARGET_NR_readlink 1035
 #define TARGET_NR_symlink 1036
 #define TARGET_NR_utimes 1037
-#define TARGET_NR_stat64 1038
-#define TARGET_NR_lstat64 1039
-#define TARGET_NR_pipe 1040
-#define TARGET_NR_dup2 1041
-#define TARGET_NR_epoll_create 1042
-#define TARGET_NR_inotify_init 1043
-#define TARGET_NR_eventfd 1044
-#define TARGET_NR_signalfd 1045
-#define TARGET_NR_sendfile 1046
-#define TARGET_NR_ftruncate 1047
-#define TARGET_NR_truncate 1048
-#define TARGET_NR_stat 1049
-#define TARGET_NR_lstat 1050
-#define TARGET_NR_fstat 1051
-#define TARGET_NR_fcntl 1052
-#define TARGET_NR_fadvise64 1053
-#define TARGET_NR_newfstatat 1054
-#define TARGET_NR_fstatfs 1055
-#define TARGET_NR_statfs 1056
-#define TARGET_NR_lseek 1057
-#define TARGET_NR_mmap 1058
+#define TARGET_NR_stat 1038
+#define TARGET_NR_lstat 1039
+
+/* FIXME: Deprecated syscalls */
 #define TARGET_NR_alarm 1059
 #define TARGET_NR_getpgrp 1060
 #define TARGET_NR_pause 1061
@@ -322,3 +305,7 @@
 #define TARGET_NR_uselib 1077
 #define TARGET_NR__sysctl 1078
 #define TARGET_NR_fork 1079
+
+/* FIXME: Just required by QEMU definitions */
+#define TARGET_NR_dup2 1041
+#define TARGET_NR_newfstatat 1054
-- 
1.7.9.5

