From c21c76d0a6820d6b20d6ecb48eb71c8bb3a9d316 Mon Sep 17 00:00:00 2001
From: Guan Xuetao <gxt@mprc.pku.edu.cn>
Date: Sat, 18 Aug 2012 02:48:34 +0800
Subject: [PATCH 154/247] unicore64-softmmu: Adjust code of switch_mode

Signed-off-by: Guan Xuetao <gxt@mprc.pku.edu.cn>
---
 target-unicore64/softmmu.c |   21 ++++++++-------------
 1 file changed, 8 insertions(+), 13 deletions(-)

diff --git a/target-unicore64/softmmu.c b/target-unicore64/softmmu.c
index 05c5f17..a73cfe8 100644
--- a/target-unicore64/softmmu.c
+++ b/target-unicore64/softmmu.c
@@ -72,24 +72,19 @@ void tlb_fill(CPUUniCore64State *env1, target_ulong addr, int is_write,
 
 void switch_mode(CPUUniCore64State *env, int mode)
 {
-    int old_mode;
     int old_mode_idx;
     int mode_idx;
 
-    old_mode = env->uncached_asr & ASR_MODE_SELECT;
-    old_mode_idx = (old_mode == ASR_MODE_PRIV) ?
+    old_mode_idx = ((env->uncached_asr & ASR_MODE_SELECT) == ASR_MODE_PRIV) ?
                     ASR_IDX_PRIV : ASR_IDX_USER;
-    mode_idx = (mode == ASR_MODE_PRIV) ?
-                    ASR_IDX_PRIV : ASR_IDX_USER;
-    if (mode_idx == old_mode_idx) {
-        return;
-    }
+    mode_idx = (mode == ASR_MODE_PRIV) ? ASR_IDX_PRIV : ASR_IDX_USER;
 
-    env->banked_r29[old_mode_idx] = env->regs[29];
-    env->banked_r30[old_mode_idx] = env->regs[30];
-
-    env->regs[29] = env->banked_r29[mode_idx];
-    env->regs[30] = env->banked_r30[mode_idx];
+    if (mode_idx != old_mode_idx) {
+        env->banked_r29[old_mode_idx] = env->regs[29];
+        env->banked_r30[old_mode_idx] = env->regs[30];
+        env->regs[29] = env->banked_r29[mode_idx];
+        env->regs[30] = env->banked_r30[mode_idx];
+    }
 }
 
 void do_interrupt(CPUUniCore64State *env)
-- 
1.7.9.5

