From de22784005ce2f93bde0d1c8f13fba4341667360 Mon Sep 17 00:00:00 2001
From: Liu Wenhui <liuwenhui@mprc.pku.edu.cn>
Date: Sat, 18 Aug 2012 19:05:41 +0800
Subject: [PATCH 167/247] unicore64-fixbug: fcmp fpcr CF

Signed-off-by: Liu Wenhui <liuwenhui@mprc.pku.edu.cn>
---
 target-unicore64/ucf64_helper.c |   34 ++++++++++++++++------------------
 1 file changed, 16 insertions(+), 18 deletions(-)

diff --git a/target-unicore64/ucf64_helper.c b/target-unicore64/ucf64_helper.c
index 06eb5bf..c0b8a70 100644
--- a/target-unicore64/ucf64_helper.c
+++ b/target-unicore64/ucf64_helper.c
@@ -228,47 +228,46 @@ void HELPER(ucf64_cmps)(float32 a, float32 b, uint32_t c,
 {
     int flag;
     flag = float32_compare_quiet(a, b, &env->ucf64.qemu_fp_status);
-    env->CF = 0;
+    env->ucf64.fpcr &= 0xd;
     switch (c & 0x7) {
     case 0: /* F */
         break;
     case 1: /* UN */
         if (flag == 2) {
-            env->CF = 1;
+            env->ucf64.fpcr |= 0x2;
         }
         break;
     case 2: /* EQ */
         if (flag == 0) {
-            env->CF = 1;
+            env->ucf64.fpcr |= 0x2;
         }
         break;
     case 3: /* UEQ */
         if ((flag == 0) || (flag == 2)) {
-            env->CF = 1;
+            env->ucf64.fpcr |= 0x2;
         }
         break;
     case 4: /* OLT */
         if (flag == -1) {
-            env->CF = 1;
+            env->ucf64.fpcr |= 0x2;
         }
         break;
     case 5: /* ULT */
         if ((flag == -1) || (flag == 2)) {
-            env->CF = 1;
+            env->ucf64.fpcr |= 0x2;
         }
         break;
     case 6: /* OLE */
         if ((flag == -1) || (flag == 0)) {
-            env->CF = 1;
+            env->ucf64.fpcr |= 0x2;
         }
         break;
     case 7: /* ULE */
         if (flag != 1) {
-            env->CF = 1;
+            env->ucf64.fpcr |= 0x2;
         }
         break;
     }
-    env->ucf64.fpcr = (env->CF << 1) | (env->ucf64.fpcr & 0xf);
 }
 
 void HELPER(ucf64_cmpd)(float64 a, float64 b, uint32_t c,
@@ -276,47 +275,46 @@ void HELPER(ucf64_cmpd)(float64 a, float64 b, uint32_t c,
 {
     int flag;
     flag = float64_compare_quiet(a, b, &env->ucf64.qemu_fp_status);
-    env->CF = 0;
+    env->ucf64.fpcr &= 0xd;
     switch (c & 0x7) {
     case 0: /* F */
         break;
     case 1: /* UN */
         if (flag == 2) {
-            env->CF = 1;
+            env->ucf64.fpcr |= 0x2;
         }
         break;
     case 2: /* EQ */
         if (flag == 0) {
-            env->CF = 1;
+            env->ucf64.fpcr |= 0x2;
         }
         break;
     case 3: /* UEQ */
         if ((flag == 0) || (flag == 2)) {
-            env->CF = 1;
+            env->ucf64.fpcr |= 0x2;
         }
         break;
     case 4: /* OLT */
         if (flag == -1) {
-            env->CF = 1;
+            env->ucf64.fpcr |= 0x2;
         }
         break;
     case 5: /* ULT */
         if ((flag == -1) || (flag == 2)) {
-            env->CF = 1;
+            env->ucf64.fpcr |= 0x2;
         }
         break;
     case 6: /* OLE */
         if ((flag == -1) || (flag == 0)) {
-            env->CF = 1;
+            env->ucf64.fpcr |= 0x2;
         }
         break;
     case 7: /* ULE */
         if (flag != 1) {
-            env->CF = 1;
+            env->ucf64.fpcr |= 0x2;
         }
         break;
     }
-    env->ucf64.fpcr = (env->CF << 1) | (env->ucf64.fpcr & 0xf);
 }
 
 /* Helper routines to perform bitwise copies between float and int.  */
-- 
1.7.9.5

