From ab0c6c6a07cc128359eddbc06dd467870d1499a2 Mon Sep 17 00:00:00 2001
From: Guan Xuetao <gxt@mprc.pku.edu.cn>
Date: Wed, 2 Jan 2013 19:53:02 +0800
Subject: [PATCH 211/247] unicore64: Rename puv4_timer to uc64_cp0_timer

Signed-off-by: Guan Xuetao <gxt@mprc.pku.edu.cn>
---
 hw/unicore64/cp0_itm.c |   22 ++++++++++------------
 1 file changed, 10 insertions(+), 12 deletions(-)

diff --git a/hw/unicore64/cp0_itm.c b/hw/unicore64/cp0_itm.c
index 4de5dc9..7361f8e 100644
--- a/hw/unicore64/cp0_itm.c
+++ b/hw/unicore64/cp0_itm.c
@@ -8,31 +8,30 @@
  * published by the Free Software Foundation, or any later version.
  * See the COPYING file in the top-level directory.
  */
+#include "cpu.h"
 #include "../ptimer.h"
 #include "../irq.h"
-#include "cpu.h"
 
 #undef DEBUG_PUV4
 #include "../puv4.h"
 
-/* puv4 ostimer implementation. */
+/* UniCore64 itimer implementation. */
 typedef struct {
     QEMUBH *bh;
     ptimer_state *ptimer;
-
 } UC64CP0State;
 
-static UC64CP0State *puv4_timer;
+static UC64CP0State *uc64_cp0_timer;
 
 uint64_t uc64_cp0_itimer_get_delta(void)
 {
-    return (uint64_t)ptimer_get_count(puv4_timer->ptimer);
+    return (uint64_t)ptimer_get_count(uc64_cp0_timer->ptimer);
 }
 
 void uc64_cp0_itimer_set_count(uint64_t value)
 {
-    ptimer_set_count(puv4_timer->ptimer, value);
-    ptimer_run(puv4_timer->ptimer, true);
+    ptimer_set_count(uc64_cp0_timer->ptimer, value);
+    ptimer_run(uc64_cp0_timer->ptimer, true);
 }
 
 static void uc64_cp0_itm_tick(void *opaque)
@@ -51,9 +50,8 @@ void uc64_cp0_itm_init(CPUUniCore64State *env)
     env->cp0.c10_itimercr = 0;
     env->cp0.c10_itimermr = 0;
 
-    puv4_timer = (UC64CP0State *)g_malloc0(sizeof(UC64CP0State));
-    puv4_timer->bh = qemu_bh_new(uc64_cp0_itm_tick, env);
-    puv4_timer->ptimer = ptimer_init(puv4_timer->bh);
-    ptimer_set_freq(puv4_timer->ptimer, 50 * 1000 * 1000);
+    uc64_cp0_timer = (UC64CP0State *)g_malloc0(sizeof(UC64CP0State));
+    uc64_cp0_timer->bh = qemu_bh_new(uc64_cp0_itm_tick, env);
+    uc64_cp0_timer->ptimer = ptimer_init(uc64_cp0_timer->bh);
+    ptimer_set_freq(uc64_cp0_timer->ptimer, 50 * 1000 * 1000);
 }
-
-- 
1.7.9.5

