From f1b4f64494fa4b954355f2eb952c419c50933969 Mon Sep 17 00:00:00 2001
From: Guan Xuetao <gxt@mprc.pku.edu.cn>
Date: Tue, 30 Aug 2011 11:41:31 +0800
Subject: [PATCH 1/8] unicore32-oldabi: add CONFIG_UNICORE32_OLDABI for
 distinguishing two ABIs

Signed-off-by: Guan Xuetao <gxt@mprc.pku.edu.cn>
---
 arch/unicore32/Kconfig | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/arch/unicore32/Kconfig b/arch/unicore32/Kconfig
index c4fbb21..4d57040 100644
--- a/arch/unicore32/Kconfig
+++ b/arch/unicore32/Kconfig
@@ -78,6 +78,17 @@ source "kernel/Kconfig.freezer"
 
 menu "System Type"
 
+config UNICORE32_OLDABI
+	bool "Allow old ABI binaries to run with this kernel"
+	default y
+	help
+	  This option preserves the old syscall interface instead of the
+	  asm-generic one. It also provides a compatibility layer to
+	  intercept syscalls that have structure arguments which layout
+	  in memory differs between the asm-generic ABI and this old one.
+
+	  If in doubt, say Y.
+
 config MMU
 	def_bool y
 
-- 
2.0.0.GIT

