From 41ae86e6c2f3927d6d08e1cd4a240359724c5d58 Mon Sep 17 00:00:00 2001
From: Guan Xuetao <gxt@mprc.pku.edu.cn>
Date: Tue, 27 Sep 2016 19:48:04 +0800
Subject: [PATCH 1/4] unicore32-oldabi: add CONFIG_UNICORE32_OLDABI for
 distinguishing two ABIs

Signed-off-by: Guan Xuetao <gxt@mprc.pku.edu.cn>
Signed-off-by: Qin Rui <qinrui@pku.edu.cn>
Tested-by: Guenter Roeck <linux@roeck-us.net>
---
 arch/unicore32/Kconfig | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/arch/unicore32/Kconfig b/arch/unicore32/Kconfig
index 0769066..e23100b 100644
--- a/arch/unicore32/Kconfig
+++ b/arch/unicore32/Kconfig
@@ -69,6 +69,17 @@ source "kernel/Kconfig.freezer"
 
 menu "System Type"
 
+config UNICORE32_OLDABI
+	bool "Allow old ABI binaries to run with this kernel"
+	default y
+	help
+	  This option preserves the old syscall interface instead of the
+	  asm-generic one. It also provides a compatibility layer to
+	  intercept syscalls that have structure arguments which layout
+	  in memory differs between the asm-generic ABI and this old one.
+
+	  If in doubt, say Y.
+
 config MMU
 	def_bool y
 
-- 
1.9.1

